rfcs/0007-custom-event-metadata-from-annotations/README.md

```diff
@@ -0,0 +1,108 @@
+# RFC-0007 Custom Event Metadata from Annotations
+
+**Status:** provisional
+
+<!--
+Status represents the current state of the RFC.
+Must be one of `provisional`, `implementable`, `implemented`, `deferred`, `rejected`, `withdrawn`, or `replaced`.
+-->
+
+**Creation date:** 2024-05-23
+
+**Last update:** 2024-05-23
+
+## Summary
+
+Flux users often run into situations where they wish to send custom, static metadata fields defined
+in Flux objects on the events dispatched by the respective controller to notification-controller.
+This proposal offers a solution for supporting those use cases uniformly across all Flux controllers
+by sending annotation keys that are prefixed with a well-defined API Group.
+
+## Motivation
+
+The need for sending static metadata in Flux objects events can benefit users in a few ways:
+
+* When sending events about identical objects living in different clusters to the same notification channel,
+users often wish to differentiate those objects by the cluster where they live in. Sending custom metadata
+fields like `cluster`, `region` and `availabilityZone` would solve this problem.
+* When implementing integrations with notification-controller, users often wish to embed pieces of information
+in Flux objects that would help them identify entities or resources that are part of their integrations, e.g.
+the ID of a Deployment in the GitHub API.
+
+### Goals
+
+Provide a method for Flux users to embed custom/static fields in their Flux objects
+and have those fields sent to notification-controller as metadata on the event payload.
+
+### Non-Goals
+
+In this proposal we **do not** aim to provide a method for Flux users to send etcd-indexed custom metadata
+fields from Flux objects in events to notification-controller, most specifically labels. By design an event
+already contains enough identification information to locate the associated Flux object inside the cluster,
+which covers the use case of labels. Flux does not wish to incentive practices that are impactful to clusters
+without a strong reason or benefit.
+
+## Proposal
+
+When sending events about Flux objects, we propose sending annotation keys prefixed with a well-defined
+API Group in addition to those already sent by the controller to implement Flux-specific functionality.
+
```

Comment:
### Example

Given this `Alert` and `HelmRelease` configuration:
```yaml
apiVersion: notification.toolkit.fluxcd.io/v1beta3
kind: Alert
metadata:
  name: slack
spec:
  providerRef:
    name: slack-bot
  eventSources:
    - kind: HelmRelease
      name: '*'
---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: podinfo
  annotations:
    event.toolkit.fluxcd.io/deployment-id: e076e315-5a48-41c3-81c8-8d8bdee7d74d
spec:
  chart:
    spec:
      chart: podinfo
      version: '6.5.*'
      sourceRef:
        kind: HelmRepository
        name: podinfo
```

The `notification-controller` would create an `Event` like this:
```json
{
  "involvedObject": {
    "apiVersion": "kustomize.toolkit.fluxcd.io/v1beta2",
    "kind": "Kustomization",
    "name": "webapp",
    "namespace": "apps",
    "uid": "7d0cdc51-ddcf-4743-b223-83ca5c699632"
  },
  "metadata": {
    "kustomize.toolkit.fluxcd.io/revision": "main/731f7eaddfb6af01cb2173e18f0f75b0ba780ef1"
    "event.toolkit.fluxcd.io/deployment-id": "e076e315-5a48-41c3-81c8-8d8bdee7d74d"
  },
  "severity":"error",
  "reason": "ValidationFailed",
  "message":"service/apps/webapp validation error: spec.type: Unsupported value: Ingress",
  "reportingController":"kustomize-controller",
  "timestamp":"2022-10-28T07:26:19Z"
}
```

Reply:
And we already do that, by the way. See this:

https://github.com/fluxcd/notification-controller/blob/3ae80f6475be1314bfecd7c69be346e16596d4a5/internal/server/event_server.go#L151-L171

Reply:
@moritzschmitz-oviva thanks for the example!